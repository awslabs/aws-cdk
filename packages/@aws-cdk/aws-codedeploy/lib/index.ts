export * from './rollback-config';
export * from './ecs';
export * from './lambda';
export * from './server';
export * from './trigger-configuration';

// AWS::CodeDeploy CloudFormation Resources:
export * from './codedeploy.generated';
