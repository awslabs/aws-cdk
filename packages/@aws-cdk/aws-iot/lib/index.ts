export * from './certificate';
export * from './policy';
export * from './policy-document';
export * from './policy-statement';
export * from './thing';

// AWS::IoT CloudFormation Resources:
export * from './iot.generated';
