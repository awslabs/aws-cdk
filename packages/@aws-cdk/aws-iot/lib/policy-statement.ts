import * as cdk from '@aws-cdk/core';

const ensureArrayOrUndefined = (field: any) => {
  if (field === undefined) {
    return undefined;
  }
  if (typeof (field) !== 'string' && !Array.isArray(field)) {
    throw new Error('Fields must be either a string or an array of strings');
  }
  if (Array.isArray(field) && !!field.find((f: any) => typeof (f) !== 'string')) {
    throw new Error('Fields must be either a string or an array of strings');
  }
  return Array.isArray(field) ? field : [field];
};

/**
 * Represents a statement in an IoT policy document.
 */
export class PolicyStatement {

  /**
   * Creates a new PolicyStatement based on the object provided.
   * This will accept an object created from the `.toJSON()` call
   * @param obj the PolicyStatement in object form.
   */
  public static fromJson(obj: any) {
    return new PolicyStatement({
      actions: ensureArrayOrUndefined(obj.Action),
      resources: ensureArrayOrUndefined(obj.Resource),
      effect: obj.Effect,
    });
  }

  /**
   * Whether to allow or deny the actions in this statement
   */
  public effect: Effect;

  private readonly action = new Array<any>();
  private readonly resource = new Array<any>();

  constructor(props: PolicyStatementProps = {}) {
    // Validate actions
    for (const action of [...props.actions || []]) {

      if (!/^(\*|[a-zA-Z0-9-]+:[a-zA-Z0-9*]+)$/.test(action) && !cdk.Token.isUnresolved(action)) {
        throw new Error(`Action '${action}' is invalid. An action string consists of a service namespace, a colon, and the name of an action. Action names can include wildcards.`);
      }
    }

    this.effect = props.effect || Effect.ALLOW;

    this.addActions(...props.actions || []);
    this.addResources(...props.resources || []);
  }

  //
  // Actions
  //

  /**
   * Specify allowed actions into the "Action" section of the policy statement.
   *
   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-policy-actions.html
   *
   * @param actions actions that will be allowed.
   */
  public addActions(...actions: string[]) {
    this.action.push(...actions);
  }

  //
  // Resources
  //

  /**
   * Specify resources that this policy statement applies into the "Resource" section of
   * this policy statement.
   *
   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-action-resources.html
   *
   * @param arns Amazon Resource Names (ARNs) of the resources that this policy statement applies to
   */
  public addResources(...arns: string[]) {
    this.resource.push(...arns);
  }

  /**
   * Adds a ``"*"`` resource to this statement.
   */
  public addAllResources() {
    this.addResources('*');
  }

  /**
   * Indicates if this permission as at least one resource associated with it.
   */
  public get hasResource() {
    return this.resource && this.resource.length > 0;
  }

  /**
   * JSON-ify the policy statement
   *
   * Used when JSON.stringify() is called
   */
  public toStatementJson(): any {
    return noUndef({
      Action: _norm(this.action),
      Effect: _norm(this.effect),
      Resource: _norm(this.resource),
    });

    function _norm(values: any) {

      if (typeof(values) === 'undefined') {
        return undefined;
      }

      if (cdk.Token.isUnresolved(values)) {
        return values;
      }

      if (Array.isArray(values)) {
        if (!values || values.length === 0) {
          return undefined;
        }

        if (values.length === 1) {
          return values[0];
        }

        return values;
      }

      if (typeof(values) === 'object') {
        if (Object.keys(values).length === 0) {
          return undefined;
        }
      }

      return values;
    }
  }

  /**
   * String representation of this policy statement
   */
  public toString() {
    return cdk.Token.asString(this, {
      displayHint: 'PolicyStatement',
    });
  }

  /**
   * JSON-ify the statement
   *
   * Used when JSON.stringify() is called
   */
  public toJSON() {
    return this.toStatementJson();
  }

  /**
   * Validate that the policy statement satisfies base requirements for a policy.
   */
  public validateForAnyPolicy(): string[] {
    const errors = new Array<string>();
    if (this.action.length === 0) {
      errors.push('An IoT PolicyStatement must specify at least one \'action\'.');
    }
    if (this.resource.length === 0) {
      errors.push('An IoT PolicyStatement must specify at least one \'resource\'.');
    }
    return errors;
  }
}

/**
 * The Effect element of an IoT policy
 *
 * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_effect.html
 */
export enum Effect {
  /**
   * Allows access to a resource in an IAM policy statement. By default, access to resources are denied.
   */
  ALLOW = 'Allow',

  /**
   * Explicitly deny access to a resource. By default, all requests are denied implicitly.
   *
   * @see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic.html
   */
  DENY = 'Deny',
}


/**
 * Interface for creating a policy statement
 */
export interface PolicyStatementProps {
  /**
   * List of actions to add to the statement
   *
   * @default - no actions
   */
  readonly actions?: string[];

  /**
   * Resource ARNs to add to the statement
   *
   * @default - no resources
   */
  readonly resources?: string[];

  /**
   * Whether to allow or deny the actions in this statement
   *
   * @default Effect.ALLOW
   */
  readonly effect?: Effect;
}

function noUndef(x: any): any {
  const ret: any = {};
  for (const [key, value] of Object.entries(x)) {
    if (value !== undefined) {
      ret[key] = value;
    }
  }
  return ret;
}
