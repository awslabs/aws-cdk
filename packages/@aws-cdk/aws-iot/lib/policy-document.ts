import * as cdk from '@aws-cdk/core';
import { PolicyStatement } from './policy-statement';

/**
 * Properties for a new PolicyDocument
 */
export interface PolicyDocumentProps {
  /**
   * Initial statements to add to the policy document
   *
   * @default - No statements
   */
  readonly statements?: PolicyStatement[];
}

/**
 * A PolicyDocument is a collection of statements
 */
export class PolicyDocument implements cdk.IResolvable {

  /**
   * Creates a new PolicyDocument based on the object provided.
   * This will accept an object created from the `.toJSON()` call
   * @param obj the PolicyDocument in object form.
   */
  public static fromJson(obj: any): PolicyDocument {
    const newPolicyDocument = new PolicyDocument();
    const statement = obj.Statement ?? [];
    if (statement && !Array.isArray(statement)) {
      throw new Error('Statement must be an array');
    }
    newPolicyDocument.addStatements(...obj.Statement.map((s: any) => PolicyStatement.fromJson(s)));
    return newPolicyDocument;
  }

  public readonly creationStack: string[];
  private readonly statements = new Array<PolicyStatement>();

  constructor(props: PolicyDocumentProps = {}) {
    this.creationStack = cdk.captureStackTrace();

    this.addStatements(...props.statements || []);
  }

  public resolve(context: cdk.IResolveContext): any {
    context.registerPostProcessor(new RemoveDuplicateStatements());
    return this.render();
  }

  /**
   * Whether the policy document contains any statements.
   */
  public get isEmpty(): boolean {
    return this.statements.length === 0;
  }

  /**
   * The number of statements already added to this policy.
   */
  public get statementCount(): number {
    return this.statements.length;
  }

  /**
   * Adds a statement to the policy document.
   *
   * @param statement the statement to add.
   */
  public addStatements(...statement: PolicyStatement[]) {
    this.statements.push(...statement);
  }

  /**
   * Encode the policy document as a string
   */
  public toString() {
    return cdk.Token.asString(this, {
      displayHint: 'PolicyDocument',
    });
  }

  /**
   * JSON-ify the document
   *
   * Used when JSON.stringify() is called
   */
  public toJSON() {
    return this.render();
  }

  /**
   * Validate that all policy statements in the policy document satisfies the
   * requirements for any policy.
   *
   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-policies.html
   */
  public validateForAnyPolicy(): string[] {
    const errors = new Array<string>();
    for (const statement of this.statements) {
      errors.push(...statement.validateForAnyPolicy());
    }
    return errors;
  }

  private render(): any {
    if (this.isEmpty) {
      return undefined;
    }

    const doc = {
      Statement: this.statements.map(s => s.toStatementJson()),
      Version: '2012-10-17',
    };

    return doc;
  }
}

/**
 * Removes duplicate statements
 */
class RemoveDuplicateStatements implements cdk.IPostProcessor {
  constructor() {
  }

  public postProcess(input: any, _context: cdk.IResolveContext): any {
    if (!input || !input.Statement) {
      return input;
    }

    const jsonStatements = new Set<string>();
    const statements: any[] = [];

    for (const statement of input.Statement) {
      const jsonStatement = JSON.stringify(statement);
      if (!jsonStatements.has(jsonStatement)) {
        statements.push(statement);
        jsonStatements.add(jsonStatement);
      }
    }

    return {
      ...input,
      Statement: statements,
    };
  }
}
