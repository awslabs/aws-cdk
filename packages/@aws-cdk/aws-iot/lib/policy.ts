import { Lazy, Resource, IResource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { ICertificate } from './certificate';
import { CfnPolicy } from './iot.generated';
import { PolicyDocument } from './policy-document';
import { PolicyStatement } from './policy-statement';
import { generatePolicyName, parsePolicyArn } from './util';
/**
 * Represents an IoT `Policy`.
 */
export interface IPolicy extends IResource {
  /**
   * The name of this IoT policy
   *
   * @attribute
   */
  readonly policyName: string;
  /**
   * The Arn of this IoT policy
   *
   * @attribute
   */
  readonly policyArn: string;
}

/**
 * Properties to initialize an instance of `Policy`.
 */
export interface PolicyProps {
  /**
   * The IoT policy name
   *
   * @default - Uses the logical ID of the policy resource, wich is ensured to
   * be unique within the stack
   */
  readonly policyName?: string;
  /**
   * The policy statements for the IoT policy
   *
   * @default - No statements.
   */
  readonly statements?: PolicyStatement[];
  /**
   * Initial PoilcyDocument to use for this Policy. If omited, any
   * `PolicyStatement` provided in the `statements` property will be applied
   * against the empty default `PolicyDocument`.
   *
   * @default - An empty policy.
   */
  readonly document?: PolicyDocument;
}
/**
 * Represents an AWS::IoT::Policy.
 *
 * AWS IoT Core policies are JSON documents. They follow the same conventions as
 * IAM policies. AWS IoT Core supports named policies so many identities can
 * reference the same policy document. Named policies are versioned so they can
 * be easily rolled back.

 * AWS IoT Core policies allow you to control access to the AWS IoT Core data
 * plane. The AWS IoT Core data plane consists of operations that allow you to
 * connect to the AWS IoT Core message broker, send and receive MQTT messages, and
 * get or update a device's shadow.

 * An AWS IoT Core policy is a JSON document that contains one or more policy
 * statements. Each statement contains:

 * - `Effect`, which specifies whether the action is allowed or denied.
 * - `Action`, which specifies the action the policy is allowing or denying.
 * - `Resource`, which specifies the resource or resources on which the action is
 * allowed or denied.
 *
 * Changes made to a policy can take a few minutes to become effective because
 * of how AWS IoT caches the policy documents. That is, it may take a few
 * minutes to access a resource that has recently been granted access, and a
 * resource may be accessisble for several minutes after its access has been
 * revoked.
 *
 * Policies can be either defined within this stack:
 *
 *   new Policy(this, 'MyPolicy', { props });
 *
 * Or imported from an existing policy:
 *
 *   Policy.import(this, 'MyImportedPolicy', { policyName: ... });
 *
 * You can also export a Policy and import it into another stack:
 *
 *   const ref = Policy.export();
 *   Policy.import(this, 'MyImportedPolicy', ref);
 *
 */
export class Policy extends Resource implements IPolicy {
  /**
   * Import a policy in this app based on it's name
   */
  public static fromPolicyName(scope: Construct, id: string, policyName: string): IPolicy {
    class Import extends Resource implements IPolicy {
      public readonly policyName = policyName;
      public readonly policyArn = parsePolicyArn(scope, policyName);
    }
    return new Import(scope, id);
  }

  /**
   * A policy document containing permissions to add to the specified resource.
   * For more information, see Access Policy Language Overview in the Amazon
   * IoT Core Developer Guide.
   */
  public readonly document = new PolicyDocument();

  private readonly _policyName: string;
  private readonly _policyArn: string;

  constructor(scope: Construct, id: string, props: PolicyProps = {}) {
    super(scope, id, {
      physicalName: props.policyName ||
        Lazy.string({ produce: () => generatePolicyName(scope, resource.logicalId) }),
    });

    if (props.document) {
      this.document = props.document;
    }

    const resource = new CfnPolicy(this, 'Resource', {
      policyName: this.physicalName,
      policyDocument: this.document,
    });

    this._policyName = this.physicalName;
    this._policyArn = resource.attrArn;

    if (props.statements) {
      props.statements.forEach(p => this.addStatements(p));
    }
  }

  /**
   * Adds a statement ot the policy document.
   */
  public addStatements(...statement: PolicyStatement[]) {
    this.document.addStatements(...statement);
  }

  /**
   * Attaches an AWS IoT policy to a principal (an x.509 certificate or other
   * credential).
   *
   * a certificate ARN (as returned from the CreateCertificate operation) or an
   * Amazon Cognito ID.
   */
  public attachToCertificate(certificate: ICertificate) {
    certificate.attachPolicy(this);
  }

  /**
   * The name of this policy
   *
   * @attribute
   */
  public get policyName(): string {
    return this._policyName;
  }

  /**
   * The Arn of this policy
   *
   * @attribute
   */
  public get policyArn(): string {
    return this._policyArn;
  }

  protected validate(): string[] {
    const errors = super.validate();
    if (this.document.isEmpty) {
      errors.push('Policy document is empty. You must add statements ot the policy');
    }
    errors.push(...this.document.validateForAnyPolicy());
    return errors;
  }
}
