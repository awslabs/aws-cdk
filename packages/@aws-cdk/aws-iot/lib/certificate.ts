import { Resource, IResource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnCertificate, CfnPolicyPrincipalAttachment, CfnThingPrincipalAttachment } from './iot.generated';
import { IPolicy } from './policy';
import { IThing } from './thing';
import { parseCertificateId, parseCertificateArn } from './util';

/**
 * All supported Certifcate status
 **/
export enum CertificateStatus {
  /**
   * Certificates must be in the ACTIVE state to authenticate devices that use a
   * certificate to connect to AWS IoT.
   */
  ACTIVE = 'ACTIVE',
  /**
   * You can create and register client certificates without activating them so
   * they can't be used until you want to use them. You can also deactivate
   * active client certificates to disable them temporarily.
   */
  INACTIVE = 'INACTIVE',
  /**
   * If you detect suspicious activity on a registered client certificate, you
   * can revoke it so that it can't be used again.
   */
  REVOKED = 'REVOKED',
  /**
   * Setting the status to PENDING_TRANSFER or PENDING_ACTIVATION will result in
   * an exception being thrown. PENDING_TRANSFER and PENDING_ACTIVATION are
   * statuses used internally by AWS IoT. They are not intended for developer
   * use.
   */
}
/**
 * All supported Certifcate modes
 **/
export enum CertificateMode {
  /**
   * Specifies which mode of certificate registration to use with this resource.
   * Valid options are DEFAULT with CaCertificatePem and CertificatePem,
   * SNI_ONLY with CertificatePem, and Default with CertificateSigningRequest.
   */
  DEFAULT = 'DEFAULT',
  /**
   * You can create and register client certificates without activating them so
   * they can't be used until you want to use them. You can also deactivate
   * active client certificates to disable them temporarily.
   */
  SNI_ONLY = 'SNI_ONLY',
}
/**
 * Represents a `Certificate`.
 */
export interface ICertificate extends IResource {
  /**
   * ARN of the IoT certificate
   * i.e. arn:aws:iot:us-east-2:123456789012:cert/certificateId
   *
   * @attribute
   */
  readonly certificateArn: string;

  /**
   * The ID of the certificate.
   * (The last part of the certificate ARN contains the certificate ID.)
   *
   * @attribute
   */
  readonly certificateId: string;

  /**
   * Adds a principal policy attachment.
   */
  attachPolicy(policy: IPolicy): void;

  /**
   * Adds a thing principal attachment.
   */
  attachThing(thing: IThing): void;
}
/**
 * Properties to initialize an instance of `Certificate`.
 */
export interface CertificateAttributes {
  /**
   * ARN of the IoT certificate
   * i.e. arn:aws:iot:us-east-2:123456789012:cert/certificateId
   *
   * @attribute
   */
  readonly certificateArn?: string;

  /**
   * The ID of the certificate.
   * (The last part of the certificate ARN contains the certificate ID.)
   *
   * @attribute
   */
  readonly certificateId?: string;
}
/**
 * Properties to initialize an instance of `Certificate`.
 */
export interface CertificateProps {
  /**
   * The status of the certificate
   *
   * @default - CertificateStatus.INACTIVE
   */
  readonly status?: CertificateStatus
  /**
   * The CA certificate used to sign the device certificate being registered,
   * not available when CertificateMode is SNI_ONLY.
   *
   * @default - none
   */
  readonly caCertificatePem?: string
  /**
   * The certificate data in PEM format. Requires SNI_ONLY for the certificate
   * mode or the accompanying CACertificatePem for registration.
   *
   * @default - none
   */
  readonly certificatePem?: string
  /**
   * Specifies which mode of certificate registration to use with this resource.
   * Valid options are DEFAULT with CaCertificatePem and CertificatePem,
   * SNI_ONLY with CertificatePem, and Default with CertificateSigningRequest.
   *
   * @default - CertificateMode.DEFAULT
   */
  readonly certificateMode?: CertificateMode
  /**
   * The certificate signing request (CSR).
   *
   * @default - none
   */
  readonly certificateSigningRequest?: string
}
/**
 * Represents an IoT Certificate.
 *
 * Certificates can be either defined within this stack:
 *
 *   new Certificate(this, 'MyCert', { props });
 *
 * Or imported from an existing certificate:
 *
 *   Certificate.import(this, 'MyImportedCertificate', { certificateArn: ... });
 *
 * You can also export a certificate and import it into another stack:
 *
 *   const ref = Certificate.export();
 *   Certificate.import(this, 'MyImportedCertificate', ref);
 *
 */
abstract class CertificateBase extends Resource implements ICertificate {
  public abstract readonly certificateArn: string;
  public abstract readonly certificateId: string;

  /**
   * Attaches an IoT Policy to the Certificate
   *
   * @param policy IPolicy
   */
  public attachPolicy(policy: IPolicy): void {
    const attachment = new CfnPolicyPrincipalAttachment(this, 'PrincipalPolicyAttachment', {
      policyName: policy.policyName,
      principal: this.certificateArn,
    });
    this.node.addDependency(policy);
    attachment.node.addDependency(policy);
  }

  /**
   * Attaches an IoT Thing to the Certificate
   *
   * @param thing IThing
   */
  public attachThing(thing: IThing): void {
    const attachment = new CfnThingPrincipalAttachment(this, 'ThingPrincipalAttachment', {
      principal: this.certificateArn,
      thingName: thing.thingName,
    });
    this.node.addDependency(thing);
    attachment.node.addDependency(thing);
  }
}

/**
 * A new Certificate
 */
export class Certificate extends CertificateBase {
  /**
   * Creates a Certificate construct that represents an external certificate.
   *
   * @param scope The parent creating construct (usually `this`).
   * @param id The construct's name.
   * @param certificateArn The Arn of the certificate to import
   */
  public static fromCertificateArn(scope: Construct, id: string, certificateArn: string): ICertificate {

    class Import extends CertificateBase {
      public readonly certificateArn = certificateArn;
      public readonly certificateId = parseCertificateId(scope, { certificateArn });
    }

    return new Import(scope, id);
  }
  /**
   * Creates a Certificate construct that represents an external certificate.
   *
   * @param scope The parent creating construct (usually `this`).
   * @param id The construct's name.
   * @param certificateId The Id of the certificate to import
   */
  public static fromCertificateId(scope: Construct, id: string, certificateId: string): ICertificate {
    class Import extends CertificateBase {
      // TODO: import certificate from another region applicable to bring your
      // own CA provisioning option?
      public readonly certificateArn = parseCertificateArn(scope, { certificateId });
      public readonly certificateId = certificateId;
    }

    return new Import(scope, id);
  }

  public readonly certificateArn: string;
  public readonly certificateId: string;

  constructor(scope: Construct, id: string, props: CertificateProps = {}) {
    super(scope, id, {});

    if (props.caCertificatePem && props.certificateMode == CertificateMode.SNI_ONLY) {
      throw new Error('Certificate invalid. CaCertificatePem not available when CertificateMode is SNI_ONLY.');
    }

    if (props.certificatePem && props.certificateMode != CertificateMode.SNI_ONLY && !props.caCertificatePem) {
      throw new Error('Certificate invalid. Requires SNI_ONLY or the accompanying CACertificatePem for registration.');
    }

    if (!props.certificateMode && !props.certificateSigningRequest && !props.certificatePem && !props.caCertificatePem) {
      throw new Error('Certificate invalid. Default requires CSR or CACertificatePem and CertificatePem');
    }

    const resource = new CfnCertificate(this, 'Resource', {
      status: props.status || CertificateStatus.INACTIVE,
      certificateMode: props.certificateMode || CertificateMode.DEFAULT,
      caCertificatePem: props.caCertificatePem,
      certificatePem: props.certificatePem,
      certificateSigningRequest: props.certificateSigningRequest,
    });

    this.certificateArn = resource.attrArn;
    this.certificateId = resource.attrId;
  }
}
