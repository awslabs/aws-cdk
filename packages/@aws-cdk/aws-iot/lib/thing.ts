import { Resource, IResource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { ICertificate } from './certificate';
import { CfnThing } from './iot.generated';
import { parseThingArn, parseThingName } from './util';
/**
 * Represents a `Thing`.
 */
export interface IThing extends IResource {
  /**
   * ARN of the IoT thing
   * i.e. arn:aws:iot:us-east-2:123456789012:thing/name
   *
   * @attribute
   */
  readonly thingArn: string;

  /**
   * The ID of the thing.
   *
   * @attribute
   */
  readonly thingName: string;

  /**
   * Attaches a certificate
   */
  attachCertificate(certificate: ICertificate): void;
}
/**
 * Properties to initialize an instance of `Thing`.
 */
export interface ThingAttributes {
  /**
   * ARN of the IoT thing
   * i.e. arn:aws:iot:us-east-2:123456789012:thing/name
   *
   * @attribute
   * @default none
   */
  readonly thingArn?: string;

  /**
   * The ID of the thing.
   *
   * @attribute
   * @default none
   */
  readonly thingName?: string;
}
/**
 * Properties to initialize an instance of `Thing`.
 */
export interface ThingProps {
  /**
   * The name of the thing
   *
   * You can't change a thing's name. To change a thing's name, you must create a
   * new thing, give it the new name, and then delete the old thing.B
   *
   * @default - generated
   */
  readonly thingName?: string
  /**
   * A string that contains up to three key value pairs. Maximum length of 800.
   * Duplicates not allowed.
   *
   * @default - none
   */
  readonly attributePayload?: CfnThing.AttributePayloadProperty
}

/**
 * Represents an IoT Thing.
 *
 * Thing can be either defined within this stack:
 *
 *   new Thing(this, 'MyCert', { props });
 *
 * Or imported from an existing Thing:
 *
 *   Thing.import(this, 'MyImportedThing', { thingArn: ... });
 *
 */
abstract class ThingBase extends Resource implements IThing {
  public abstract readonly thingArn: string;
  public abstract readonly thingName: string;

  /**
   * Attaches an IoT Policy to the Thing
   *
   * @param certificate ICertificate
   */
  public attachCertificate(certificate: ICertificate): void {
    certificate.attachThing(this);
  }
}

/**
 * A new Thing
 */
export class Thing extends ThingBase {
  /**
   * Creates a thing construct that represents an external Thing.
   *
   * @param scope The parent creating construct (usually `this`).
   * @param id The construct's name.
   * @param thingArn The Arn of the thing to import
   */
  public static fromThingArn(scope: Construct, id: string, thingArn: string): IThing {

    class Import extends ThingBase {
      public readonly thingArn = thingArn;
      public readonly thingName = parseThingName(scope, { thingArn });
    }

    return new Import(scope, id);
  }
  /**
   * Creates a Thing construct that represents an external thing.
   *
   * @param scope The parent creating construct (usually `this`).
   * @param id The construct's name.
   * @param thingName The Id of the thing to import
   */
  public static fromThingName(scope: Construct, id: string, thingName: string): IThing {
    class Import extends ThingBase {
      public readonly thingArn = parseThingArn(scope, { thingName });
      public readonly thingName = thingName;
    }

    return new Import(scope, id);
  }

  public readonly thingArn: string;
  public readonly thingName: string;

  constructor(scope: Construct, id: string, props: ThingProps = {}) {
    super(scope, id, {
      physicalName: props.thingName,
    });

    if (props && props.attributePayload && props.attributePayload.attributes) {
      if (Object.keys(props.attributePayload?.attributes).length > 3) {
        throw new Error('Invalid thing attributes. Can only have three attributes');
      }
      if (JSON.stringify(props.attributePayload?.attributes).length > 800) {
        throw new Error('Invalid thing attribute length. Can not exceed 800 characters');
      }
    }

    const resource = new CfnThing(this, 'Thing', {
      thingName: this.physicalName,
      attributePayload: props.attributePayload,
    });

    this.thingName = resource.logicalId;
    this.thingArn = resource.getAtt('Arn').toString();
  }
}
