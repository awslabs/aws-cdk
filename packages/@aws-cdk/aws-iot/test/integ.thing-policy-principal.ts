import { App, Stack, Arn } from '@aws-cdk/core';
import { Certificate, PolicyStatement, Policy, Thing } from '../lib';


const csr = `-----BEGIN CERTIFICATE REQUEST-----
MIICrTCCAZUCAQAwaDELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAklMMRAwDgYDVQQH
DAdDaGljYWdvMRUwEwYDVQQDDAxkdW1teS1kZXZpY2UxIzAhBgkqhkiG9w0BCQEW
FG1lQGRhcnJlbmhvbGxhbmQuY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIB
CgKCAQEAmjJYlBrXBg6fDZStT30J8j0iMaDSYc1+RUa12ECbKDVYtjVH2rvG4oDF
9Mz89C37wzvaNmKZZRNS/pVAspZr6EII68ZSuPe52USNo48rUGwtS9gvk61CP3Fo
IAhiWpbi0F2waP6g+XZ/qFxOdqTKkEAhJ087ozSGsf7ZoHqB63OaapK8SNOB5bMf
mf3o3rn0xLpBq50ujevaiJVIiJ9fR2ZBDb2O7VSZn6YQOHZhef7lLgQhrzuIXR/d
ieXyo/38nnFt+pLtfN7RUauZX0J1EDfBDcgicjP+Z5Fwz/PN7cwq1ZllNkVXSGb7
Mg82MSTCBiLd4O1MFWZGdkhblN8/RQIDAQABoAAwDQYJKoZIhvcNAQELBQADggEB
ADy/BLvx+twEsIO8izl5jkf3oNoCSs8FaU3Q9iy8zgVYMxLrNzcYlGxFLAZhfgMh
cbCm9l0XcZDig6MLxVdu04hcIfunDRNpfYpbQGbTGanivf4UYz0sokrXK12Vy5yS
3etMX8Eakd7fwiCLly8ulRGYvSKIcK/XQzv3Kz/AnK0JxCXyFhDuBHKwAsdH0ipv
xcoCaNawiGQs2vG1m9CE+NZZ9gcvMAFNTT5wC2D/RpqO0cBdDDBAOwec1nAub8Jb
TzX8iklDwVoWn16qSfOxzcCpHEytlZRwVBsmqNhlRDA8+Q/LByLC9/mDlOxwiF4V
//xfxL6/JkoTKIHNUP7Hl6E=
-----END CERTIFICATE REQUEST-----`;

const app = new App();
const stack = new Stack(app, 'aws-cdk-thing-policy-principal');

// Use CSR to register certifiate. This will surely break tests
const cert = new Certificate(stack, 'MyThingCertificate', {
  certificateSigningRequest: csr,
});

// Create thing
const thing = new Thing(stack, 'MyThing');

// Create policy and add thing dependeny for attachment
const policy = new Policy(stack, 'MyThingPolicy');
policy.node.addDependency(thing);

// Allow IoT Connect
const connect = new PolicyStatement({
  actions: ['iot:Connect'],
});
connect.addAllResources();

// Allow pub/sub to topics and topicfilters provided to resources
const pubsub = new PolicyStatement({
  actions: [
    'iot:Receive',
    'iot:Subscribe',
    'iot:Publish',
  ],
});

// Provide thing as resource
pubsub.addResources(Arn.format({
  resource: 'topic',
  service: 'iot',
  resourceName: thing.thingName,
}, stack));

// Add statements to policy
policy.addStatements(connect, pubsub);

// Add policy to certificate
cert.attachPolicy(policy);

// Attach certificate to thing
// thing.attachCertificate(cert);
