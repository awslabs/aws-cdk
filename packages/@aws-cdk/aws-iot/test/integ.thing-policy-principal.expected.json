{
  "Resources": {
    "MyThingCertificateC4014F8E": {
      "Type": "AWS::IoT::Certificate",
      "Properties": {
        "Status": "INACTIVE",
        "CertificateMode": "DEFAULT",
        "CertificateSigningRequest": "-----BEGIN CERTIFICATE REQUEST-----\nMIICrTCCAZUCAQAwaDELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAklMMRAwDgYDVQQH\nDAdDaGljYWdvMRUwEwYDVQQDDAxkdW1teS1kZXZpY2UxIzAhBgkqhkiG9w0BCQEW\nFG1lQGRhcnJlbmhvbGxhbmQuY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIB\nCgKCAQEAmjJYlBrXBg6fDZStT30J8j0iMaDSYc1+RUa12ECbKDVYtjVH2rvG4oDF\n9Mz89C37wzvaNmKZZRNS/pVAspZr6EII68ZSuPe52USNo48rUGwtS9gvk61CP3Fo\nIAhiWpbi0F2waP6g+XZ/qFxOdqTKkEAhJ087ozSGsf7ZoHqB63OaapK8SNOB5bMf\nmf3o3rn0xLpBq50ujevaiJVIiJ9fR2ZBDb2O7VSZn6YQOHZhef7lLgQhrzuIXR/d\nieXyo/38nnFt+pLtfN7RUauZX0J1EDfBDcgicjP+Z5Fwz/PN7cwq1ZllNkVXSGb7\nMg82MSTCBiLd4O1MFWZGdkhblN8/RQIDAQABoAAwDQYJKoZIhvcNAQELBQADggEB\nADy/BLvx+twEsIO8izl5jkf3oNoCSs8FaU3Q9iy8zgVYMxLrNzcYlGxFLAZhfgMh\ncbCm9l0XcZDig6MLxVdu04hcIfunDRNpfYpbQGbTGanivf4UYz0sokrXK12Vy5yS\n3etMX8Eakd7fwiCLly8ulRGYvSKIcK/XQzv3Kz/AnK0JxCXyFhDuBHKwAsdH0ipv\nxcoCaNawiGQs2vG1m9CE+NZZ9gcvMAFNTT5wC2D/RpqO0cBdDDBAOwec1nAub8Jb\nTzX8iklDwVoWn16qSfOxzcCpHEytlZRwVBsmqNhlRDA8+Q/LByLC9/mDlOxwiF4V\n//xfxL6/JkoTKIHNUP7Hl6E=\n-----END CERTIFICATE REQUEST-----"
      },
      "DependsOn": [
        "MyThingPolicyABA3B7A8"
      ]
    },
    "MyThingCertificatePrincipalPolicyAttachmentD23C5737": {
      "Type": "AWS::IoT::PolicyPrincipalAttachment",
      "Properties": {
        "PolicyName": "MyThingPolicyABA3B7A8",
        "Principal": {
          "Fn::GetAtt": [
            "MyThingCertificateC4014F8E",
            "Arn"
          ]
        }
      },
      "DependsOn": [
        "MyThingPolicyABA3B7A8"
      ]
    },
    "MyThing0C5333DD": {
      "Type": "AWS::IoT::Thing"
    },
    "MyThingPolicyABA3B7A8": {
      "Type": "AWS::IoT::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iot:Connect",
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": [
                "iot:Receive",
                "iot:Subscribe",
                "iot:Publish"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":iot:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":topic/MyThing0C5333DD"
                  ]
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyThingPolicyABA3B7A8"
      },
      "DependsOn": [
        "MyThing0C5333DD"
      ]
    }
  }
}